lgc/interface/lgc/ElfLinker.h

```diff
@@ -0,0 +1,84 @@
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  ElfLinker.h
+ * @brief LLPC header file: LGC interface for linking unlinked half-pipeline ELFs into pipeline ELF
+ ***********************************************************************************************************************
+ */
+#pragma once
+
+#include "lgc/CommonDefs.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace lgc {
+
+// =====================================================================================================================
+// The public API of the LGC interface for ELF linking.
+// The ElfLinker object is created by calling Pipeline::getElfLinker(). The ElfLinker internally refers back to
+// its Pipeline, and thus uses pipeline state for glue code generation, adding to PAL metadata, and resolving
+// relocs.
+class ElfLinker {
+public:
+  // Get information on the glue code that will be needed for the link. It is an implementation detail how many
+  // chunks of glue there might be and what they are for, but, for information, they will be some subset of:
+  // - A CS prolog
+  // - A VS prolog ("fetch shader")
+  // - A vertex-processing epilog ("parameter export shader")
+  // - An FS epilog ("color export shader")
+  //
+  // Returns an array (possibly 0 length) with an entry for each chunk of glue code, where an entry
+  // is a StringRef that the client can hash for its cache lookup. If it gets a cache hit, it should provide the
+  // found blob to ElfLinker::addGlue. If it does not get a cache hit, the client can call ElfLinker::compileGlue to
+  // retrieve the compiled glue code to store in the cache.
+  virtual llvm::ArrayRef<llvm::StringRef> getGlueInfo() = 0;
```

Comment:
Should this just be a hash then instead of a StringRef?

Reply:
I did wonder about that, but that ties the front-end (and/or driver) into a particular size of hash that might disagree with other stuff it does. So I thought it would be better to let the front-end do the hashing in the same way it does other hashing.

Reply:
Fair enough.